crates/bevy_app/src/app.rs

```diff
@@ -288,6 +288,16 @@ impl App {
             panic!("App::run() was called from within Plugin::build(), which is not allowed.");
         }
 
+        // Finish the startup phase and run one clean pass of the main schedule.
+        Main::startup(&mut app.world);
+        Main::run_main(&mut app.world);
+
+        // Force plugins to finish their setup and advance by one frame to make sure everything is
+        // setup up. This is important to prevent black frames during the launch animation on iOS,
+        // but it seems to also solve a brief flicker during startup on Linux/Wayland.
+        app.finish();
```

Comment:
you can't call `app.finish()` until `app.ready()` will return true

Reply:
Ah, I may have misunderstood then. I thought `finish()` would run the plugin initialization to completion so that `ready()` will return `true` afterwards.

Then what would be the recommended way to wait for `ready()` to return true?

Reply:
Bevy waits for `ready()` by starting the winit event loop and running it until it's true, which is probably what you're trying to work around on iOS

Reply:
I see. You’re right I must work around that though, because once control is handed over to the event loop it means iOS’ `didFinishLaunching` returns and the transition starts. Everything must be ready synchronously or I’m in trouble and the black frames are back.

It makes me wonder though, does that mean some plugins *need* an event loop in order to become ready? If yes, is it just a few? I suppose, since I didn’t run into issues on iOS and Linux. And would it be possible to adjust those plugins?

Or if no, why start the event loop at all that before plugins are ready?

Reply:
This is mandatory for webgpu support, for renderer initialization, and was done on all platforms to keep it the same.

In wgpu, the renderer init is async on all platforms, but the actual duration varies. I guess on iOS and Linux (and maybe then, not on all devices) it can be fast enough

Reply:
could you try https://github.com/bevyengine/bevy/pull/9830?

this PR should make `ready` return `true` directly on native, so you'll be able to do something like
```rust
if app.ready() {
  .. your code changes
}
```

Reply:
Thanks, that's very helpful! I will give it a shot tonight!

Reply:
Thanks, that patch works fine with this PR! I've cherry-picked it into this one too to see if those suggestions together will also fix the CI.

Reply:
CI is still red, but I just tested it, and it appears that with your change it's enough to only call `.finish()` and I don't need to force `startup()` anymore, so I'll indeed revert that part.